### PR TITLE
Implementing DataOps with Snowflake and AWS

### PR DESCRIPTION
**Pull request by zmaleki**

### CODE REVIEW

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+<center><img src="/img/blog/dataops/secretsmanager-1.png" /></center><br/>
+Make sure to name it "snowflake/pipeline_sys_user"
+<center><img src="/img/blog/dataops/secretsmanager-2.png" /></center><br/>
+
+Last step is allowing code build in your pipeline to access the key in secrets manager.
+To do that find out the arn of your secrets manager:
+<center><img src="/img/blog/dataops/secretsmanager-3.png" /></center><br/>
+Open aws_seed.yaml file and update line number 404 with the arn to your secret manager password
+```
+  - <The arn to the secrets manager that holds Snowflake Password>
```
Comment (paulmech) [inline _posts/2020-10-30-snowflake-dataOps.md:64]: couldn't this be a parameter in the cloudformation template?

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+CREATE WAREHOUSE pipeline_warehouse;
+GRANT USAGE ON WAREHOUSE pipeline_warehouse TO ROLE pipeline_role;
+
+// Ideally you only want to grant permissions that your pipeline requires. Granting SYSADMIN is not encouraged
+grant role SYSADMIN to role pipeline_role;
+
+```
+
+Optional:
+In my sql scripts, I am creating tables in my_db database and my_schema schema. So if you want my template scripts run seccessfully on your snowflake account, you will need to run below statements too
```
Comment (paulmech) [inline _posts/2020-10-30-snowflake-dataOps.md:86]: ~seccessfully~ successfully

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    "ParameterKey": "SnowflakeRole",
+    "ParameterValue": "pipeline_role"
+  }
+```
+
+# Step 4: Deploy your code to AWS
+To deploy the infrastructure for your pipeline, you will need to first setup your aws credentials in your terminal. Next you will need to execute init.sh file
+
+```
+sh pipeline/init.sh
```
Comment (paulmech) [inline _posts/2020-10-30-snowflake-dataOps.md:134]: could be worth noting that the aws user you are running as will need admin-like privileges, e.g. be able to create iam roles etc.

**Comment by paulmech** [comment]
@zmaleki Great job, cool article and brilliant to cover Flyway instead of sqitch as many people will be much more familiar with Flyway I think.

I think it would be good to expand on what dataops is, and how devops is just one part of it. For example you could link to and  reflect on the principles in [dataops manifesto](https://www.dataopsmanifesto.org/) or [this article](https://medium.com/data-ops/dataops-is-not-just-devops-for-data-6e03083157b7) to highlight how dataops favours being able to test and verify data transformations for example.

In your intro section you write

`In this blog post, I will show you how to create your Snowflake DataOps pipeline using AWS developer tools and Flyway.`

but you could expand that to say things like
- in Snowflake just about everything can be managed via executable SQL statements, this makes schema migration tools like Flyway and sqitch the perfect devops companion tool for Snowflake
- It is great for creating and managing resources, like grants, databases resources, snowpipes, stages, tasks, integrations etc
- While it is possible to manage data transformations (e.g. using tasks) using flyway, this is a job better suited to tools like dbt

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+Make sure to follow Flyway the naming conventions. All version sql files should start with V__ and repeatable scripts should start with R___
+<img src="/img/blog/dataops/flyway.png" /><br/>
+
+
+# References:
+
+Flyway: 
+https://flywaydb.org/documentation/database/snowflake
+
+Inception Pipeline:
```
Comment (paulmech) [inline _posts/2020-10-30-snowflake-dataOps.md:156]: I'd remove these links for Inception Pipeline or include only one of them

File: _posts/2020-10-30-snowflake-dataOps.md (hunk 1 above)
Comment (zmaleki) [reply _posts/2020-10-30-snowflake-dataOps.md:64]: Because the secret is stored manually it can't be used as a export parameter

Comment (zmaleki) [reply _posts/2020-10-30-snowflake-dataOps.md:64]: I'll change it to pass it as parameter but they still have to enter the arn manually

**Review by Hamo69 — APPROVED** [review]
Love this blog, I can see it going viral

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+tags: devops, dataops, aws, snowflake, ci, cd, flyway
+author: Zainab Maleki
+image: img/blog/dataops/dataOps.jpg
+---
+
+<center><img src="/img/blog/dataops/dataOps.jpg" /></center><br/>
+
+# What is DataOps?
+
+DevOps practices have been known, utilised and perfected over the past few years to allow speed and agility in the development life cycles. While all the advancements have happened in development lands, data domains have not kept pace. DataOps is a concept that allows applying DevOps practices to data life cycles to not only achieve speed and agility but also to deliver trusted and high-quality data.<br/>
```
Comment (matt-tyler) [inline _posts/2020-10-30-snowflake-dataOps.md:14]: the negative 'to not only achieve is' difficult to comprehend while skimming, and there are a few grammar problems as well.

`DevOps practices have been discovered, utilised and perfected over the past few years to allow speed and agility in software engineering. Despite these advancements in the software engineering discipline, they have not been widely adopted in the data engineering domain. DataOps is a concept that applies DevOps practices to data life cycles to achieve speed and agility whilst delivering trusted and high-quality data.`

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+* CodeCommit
+* CodeBuild
+* CodePipeline
+* Cloudformation
+* Flyway
+
+# Step 2: Setup Snowflake Credentials
+
+To allow your pipeline to get access to Snowflake, you will need to first create a RSA public and private key. You can create the keys using openssl by running below commands in your command line
```
Comment (matt-tyler) [inline _posts/2020-10-30-snowflake-dataOps.md:39]: `an RSA public`

File: _posts/2020-10-30-snowflake-dataOps.md (hunk 4 above)
Comment (matt-tyler) [reply _posts/2020-10-30-snowflake-dataOps.md:156]: We really should stop using the inception pipeline for small guides - it's an incredible amount of content to consume just to kick the tyres on something else.

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+* CodeCommit
+* CodeBuild
+* CodePipeline
+* Cloudformation
+* Flyway
+
+# Step 2: Setup Snowflake Credentials
+
+To allow your pipeline to get access to Snowflake, you will need to first create an RSA public and private key. You can create the keys using openssl by running below commands in your command line
+```
```
Comment (sbracegirdle) [inline _posts/2020-10-30-snowflake-dataOps.md:40]: tip; you can get language specific code formatting by putting the language code after the three ticks.

e.g. \`\`\`sql and \`\`\`sh

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+tags: devops, dataops, aws, snowflake, ci, cd, flyway
+author: Zainab Maleki
+image: img/blog/dataops/dataOps.jpg
+---
+
+<center><img src="/img/blog/dataops/dataOps.jpg" /></center><br/>
+
+# What is DataOps?
+
+DevOps practices have been discovered, utilised and perfected over the past few years to allow speed and agility in software engineering. Despite these advancements in the software engineering discipline, they have not been widely adopted in the data engineering domain. DataOps is a concept that applies DevOps practices to data life cycles to achieve speed and agility whilst delivering trusted and high-quality data. Like DevOps, DataOps minimises manual interventions by codifying data related changes and deployment through automated CI/CD tools.<br/><br/>
```
Comment (sbracegirdle) [inline _posts/2020-10-30-snowflake-dataOps.md:14]: Minor; I don't think you need the line breaks (i.e. `<br/>`). If there's a space between lines I think it treats them as separate paragraphs and puts a margin in between

File: _posts/2020-10-30-snowflake-dataOps.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+2. Pushing your codebase into the newly created Codecommit repository
+
+If the second step fails because of your git config settings, you will need to make sure to run them again
+<center><img src="/img/blog/dataops/init.png" /></center><br/>
+Now you can check codepipeline and see a green end to end deployment
+<left><img src="/img/blog/dataops/codepipeline.png" /></left><br/>
+
+This repo uses Flyway to deploy Snowflake changes. To deploy new changes, all you need is placing your new sql scripts in the sql folder and pushing it to codecommit.
+
+Make sure to follow Flyway the naming conventions. All version sql files should start with V__ and repeatable scripts should start with R___
```
Comment (sbracegirdle) [inline _posts/2020-10-30-snowflake-dataOps.md:150]: Suggestion; maybe add an example here to demonstrate the naming convention and folder structure?

**Review by sbracegirdle — APPROVED** [review]

File: _posts/2020-10-30-snowflake-dataOps.md (hunk 9 above)
Comment (zmaleki) [reply _posts/2020-10-30-snowflake-dataOps.md:150]: I have added a photo example of my files

**Pull request merged**